### ISSUE TITLE
rgbds 0.4.2-pre `startCapture` cannot capture part of a string expansion

### ISSUE THREAD

**Issue by Rangi42** [gbdev/rgbds#590]
This builds to just `20 20` with rgbds 0.4.1:

```
SECTION "test", ROM0[0]

FOO EQUS "\ndouble: MACRO\nREPT 2\n\tdb \\1\nENDR\nENDM"
	FOO
	PURGE FOO

	double $20
```

But with 0.4.2-pre it recurses infinitely (and has wrong line numbers):

```
ERROR: s.asm(11) -> s.asm::double(5):
    'FOO' not defined
ERROR: s.asm(11) -> s.asm::double(7) -> s.asm::double(5):
    'FOO' not defined
[...]
FATAL: s.asm(11) -> s.asm::double(7) -> [...] -> s.asm::double(7):
    Recursion limit (64) exceeded
```

**Comment by Rangi42** [comment]
Without the `REPT` it's just a `syntax error`:

```
SECTION "test", ROM0[0]

FOO EQUS "\ndouble: MACRO\n\tdb \\1, \\1\nENDM"
	FOO
	PURGE FOO

	double $20
```

**Comment by Rangi42** [comment]
Defining `LEXER_DEBUG` reveals that `lexer_CaptureMacroBody` is passing to `sym_AddMacro` the wrong `body` value. It's the text in the actual file after the usage of `FOO`, not the text within the expansion of `FOO`.

**Comment by Rangi42** [comment]
So this isn't just a `MACRO` issue. It applies to other control structures.

In 0.4.1, this outputs the ROM `2a 2a 2a`:

```
SECTION "test", ROM0[0]

BAR EQUS "\nREPT 3\n\tdb $2a\nENDR"

	BAR
```

But in 0.4.2-pre it fails with various errors depending on what (if anything) comes after the invocation `BAR`.

**Comment by Rangi42** [comment]
It's tempting to suggest just banning `REPT`, `ENDR`, `MACRO`, and `ENDM` from being inside an `EQUS` expansion. That would be the easiest solution, and probably nobody is depending on that behavior. However, it would be inconsistent with all the other things that *do* work inside an `EQUS`.

"`ENDC`" can go in an `EQUS`: this is a valid file that outputs the ROM `42` in both 0.4.1 and 0.4.2-pre:

```
SECTION "test", ROM0[0]
BAR EQUS " db $42\n\
ENDC"
IF 1
 BAR 
```

So can "`SECTION`": this outputs `01 02 03 04 00 00 00 00 05 06 07 08`:

```
SECTION "test", ROM0[0]
BAR EQUS "SECTION \"test2\", ROM0[8]"
 db 1, 2, 3, 4
 BAR
 db 5, 6, 7, 8
```

So can `PURGE`, which is essential for purging symbols via a symbol name stored in an `EQUS` string; see https://github.com/gbdev/rgbds/issues/269#issuecomment-684114877 . I expect that basically every rgbds keyword works, with these two `startCapture`-dependent cases being the current (in 0.4.2-pre) exception.

**Comment by ISSOtm** [comment]
I should check, but I think that `startCapture` assumes that an expansion cannot start within an expansion, and this is why the bug occurs.

Since `EQUS` are not expanded when scanning for a macro, this means that the only `EQUS` a macro definition can contain would be at its start; if so, the fix should be to check if the macro begins in an `EQUS`, and if so, capture on the non-`mmap` path, in order to both avoid lifetime problems if the underlying `EQUS` is deleted, and properly handle a macro whose definition *begins* in an `EQUS`, but *ends* outside of one.

It's worth considering whether we want to allow that last case, though: it doesn't seem to be too bad, but I'm worried it might be confusing that an `EQUS` is only expanded at its beginning, but still fully expanded... what?

**Comment by ISSOtm** [comment]
It was as I guessed: `rgbasm: src/asm/lexer.c:1965: startCapture: Assertion `!lexerState->expansions' failed.`

I'm going to keep the current behavior for now, and maybe deprecate it later. For now, I want to fix the last two blockers and release 0.4.2.
